src/platforms/stm32/src/lib/gpio_driver.c

```diff
@@ -229,6 +229,19 @@ static term create_pair(Context *ctx, term term1, term term2)
     return ret;
 }
 
+static term create_error_tuple(Context *ctx, term err_atom)
+{
+    if (UNLIKELY(!term_is_atom(err_atom))) {
+        AVM_LOGE(TAG, "Error term 'err_atom' was not an atom, returning 'undefined'.");
+        err_atom = UNDEFINED_ATOM;
+    }
+    if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
+
+    return create_pair(ctx, ERROR_ATOM, err_atom);
```

Comment:
I noticed that there are still calls to `create_pair` (in `consume_gpio_mailbox`) that do not do memory_ensure_free's beforehand.  So you might want to check all usages to make sure you are not allocating a term in the heap without first checking.

In general, I try to shy away from doing the check in a common function like this, but I understand why -- it makes a code a lot simpler, for sure.  I don't have an easy solution.

Reply:
Maybe a name change for `create_error_tuple` would be better. Something like `create_error_tuple_maybe_gc` or `safe_create_error_tuple`? Or is that just trying to put lipstick on a pig?😂

Reply:
I fixed the `create_pair` calls that were unsafe, and opted to change the name of `create_error_tuple` to `reason_atom_to_error_tuple_safely`, not at all elegant, but I think this eliminates a lot of future confusion.